GH-5291 Add asynchronous fsync to LuceneSail

Ostrzyciel opened the pull request:
GitHub issue resolved: #5291

Briefly describe the changes proposed in this PR:

As described in the issue, the current setup with an fsync after each transaction is very safe, but also a huge bottleneck when dealing with many small transactions. This PR introduces an option that allows for asynchronous fsyncs in the background, on a fixed interval. If there is nothing to sync, it does nothing.

I tested this with the original workload with which I found the issue. When I set `fsyncInterval` to 5000 ms, it went from ~10–12 TX/s to ~100–150 TX/s, over an HTTP connection. That's basically the same as when I tried removing the fsync entirely. Great :)

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

hmottestad commented:
I don't think that the close() method on the directory in the LuceneIndex class is ever called.

Ostrzyciel commented:
> I don't think that the close() method on the directory in the LuceneIndex class is ever called.

@hmottestad oops, you are right! Fixed that and added a test to make sure it happens.

hmottestad commented:
Thanks for the good fix. I think this will be a good solution overall, just some small things I want to be sure are robust.

Ostrzyciel commented on core/sail/lucene/src/main/java/org/eclipse/rdf4j/sail/lucene/impl/DelayedSyncDirectoryWrapper.java:
```diff
@@ -0,0 +1,103 @@
+	}
+
+	@Override
+	public void close() throws IOException {
+		// Finish the current sync task, if in progress and then shut down
+		scheduler.shutdown();
+		// Do a final sync of any remaining files
+		doSync();
+		super.close();
+	}
```
Done.

Ostrzyciel commented:
> @Ostrzyciel if it gets too complicated I'm open for adding a system property for disabling fsync in the lucene sail, if you believe that's a viable option for you.

Yeah, it would work for my case fine. I'm wrapping up the implementation of this right now, running the end-to-end tests.
